sql/database.go

```diff
@@ -818,13 +818,13 @@ func (db *sqliteDatabase) WithConnection(ctx context.Context, exec func(Executor
 	}
 	conCtx, cancel := context.WithCancel(ctx)
 	conn := db.getConn(conCtx)
+	if conn == nil {
+		return ErrNoConnection
+	}
 	defer func() {
 		cancel()
 		db.pool.Put(conn)
 	}()
-	if conn == nil {
-		return ErrNoConnection
-	}
```

Comment:
This leaks a go routine, you need to ensure that `cancel` is called after creating `conCtx`. How about:

```go
	conCtx, cancel := context.WithCancel(ctx)
	defer cancel()
	conn := db.getConn(conCtx)
	if conn == nil {
		return ErrNoConnection
	}
	defer db.pool.Put(conn)
```

Reply:
Thanks, fixed